service/storage_proxy.cc

```diff
@@ -604,102 +604,110 @@ class storage_proxy::remote {
+        query::read_command cmd1, ::compat::wrapping_partition_range pr,
+        rpc::optional<query::digest_algorithm> oda,
+        rpc::optional<db::per_partition_rate_limit::info> rate_limit_info_opt)
+    {
         tracing::trace_state_ptr trace_state_ptr;
         auto src_addr = netw::messaging_service::get_source(cinfo);
         if (cmd1.trace_info) {
             trace_state_ptr = tracing::tracing::get_local_tracing_instance().create_session(*cmd1.trace_info);
             tracing::begin(trace_state_ptr);
-            tracing::trace(trace_state_ptr, "read_data: message received from /{}", src_addr.addr);
+            tracing::trace(trace_state_ptr, "{}: message received from /{}", verb, src_addr.addr);
         }
-        auto da = oda.value_or(query::digest_algorithm::MD5);
         auto rate_limit_info = rate_limit_info_opt.value_or(std::monostate());
         if (!cmd1.max_result_size) {
-            auto& cfg = _sp.local_db().get_config();
-            cmd1.max_result_size.emplace(cfg.max_memory_for_unlimited_query_soft_limit(), cfg.max_memory_for_unlimited_query_hard_limit());
+            if constexpr (verb == read_verb::read_data) {
+                auto& cfg = _sp.local_db().get_config();
+                cmd1.max_result_size.emplace(cfg.max_memory_for_unlimited_query_soft_limit(), cfg.max_memory_for_unlimited_query_hard_limit());
+            } else {
+                cmd1.max_result_size.emplace(cinfo.retrieve_auxiliary<uint64_t>("max_result_size"));
+            }
         }
-
         shared_ptr<storage_proxy> p = _sp.shared_from_this();
         auto cmd = make_lw_shared<query::read_command>(std::move(cmd1));
-        p->get_stats().replica_data_reads++;
         auto src_ip = src_addr.addr;
         auto timeout = t ? *t : db::no_timeout;
         schema_ptr s = co_await get_schema_for_read(cmd->schema_version, std::move(src_addr), timeout);
         auto pr2 = ::compat::unwrap(std::move(pr), *s);
-        if (pr2.second) {
-            // this function assumes singular queries but doesn't validate
-            throw std::runtime_error("READ_DATA called with wrapping range");
-        }
-        query::result_options opts;
-        opts.digest_algo = da;
-        opts.request = da == query::digest_algorithm::none ? query::result_request::only_result : query::result_request::result_and_digest;
-        future<rpc::tuple<foreign_ptr<lw_shared_ptr<query::result>>, cache_temperature>> f = co_await coroutine::as_future(p->query_result_local(std::move(s), cmd, std::move(pr2.first), opts, trace_state_ptr, timeout, rate_limit_info));
-        tracing::trace(trace_state_ptr, "read_data handling is done, sending a response to /{}", src_ip);
-        co_return co_await encode_replica_exception_for_rpc<rpc::tuple<foreign_ptr<lw_shared_ptr<query::result>>, cache_temperature, replica::exception_variant>>(p->features(), std::move(f));
+        auto f = co_await coroutine::as_future(std::invoke([&]() {
```

Comment:
Did this introduce a new allocation?

Better to move the template to the calling scope if so.

Reply:
There were the same `coroutine::as_future` in the previous version. What allocation do you mean?

Reply:
Every coroutine introduces an allocation. But there's net zero coroutine changes, so we're fine.